src/twisted/internet/asyncioreactor.py

```diff
@@ -9,42 +9,56 @@
 
 import errno
 
+from typing import Dict, Optional, Type, cast
+
 from zope.interface import implementer
 
 from twisted.logger import Logger
 from twisted.internet.posixbase import (PosixReactorBase, _NO_FILEDESC,
                                         _ContinuousPolling)
 from twisted.python.log import callWithLogger
 from twisted.python.runtime import seconds as runtimeSeconds
+from twisted.internet.abstract import FileDescriptor
 from twisted.internet.interfaces import IReactorFDSet
 
-try:
-    from asyncio import get_event_loop
-except ImportError:
-    raise ImportError("Requires asyncio.")
-
-# As per ImportError above, this module is never imported on python 2, but
-# pyflakes still runs on python 2, so let's tell it where the errors come from.
-from builtins import PermissionError, BrokenPipeError
+from asyncio import get_event_loop, AbstractEventLoop, SelectorEventLoop
 
 
 
 @implementer(IReactorFDSet)
 class AsyncioSelectorReactor(PosixReactorBase):
     """
     Reactor running on top of L{asyncio.SelectorEventLoop}.
+
+    On POSIX platforms, the default event loop is
+    L{asyncio.SelectorEventLoop}.
+    On Windows, the default event loop on Python 3.7 and older
+    is L{asyncio.WindowsSelectorEventLoop}, but on Python 3.8 and newer
+    the default event loop is L{asyncio.WindowsProactorEventLoop} which
+    is incompatible with L{AsyncioSelectorReactor}.
+    Applications that use L{AsyncioSelectorReactor} on Windows
+    with Python 3.8+ must call
+    C{asyncio.set_event_loop_policy(asyncio.WindowsSelectorEventLoopPolicy())}
+    before instantiating and running L{AsyncioSelectorReactor}.
     """
     _asyncClosed = False
     _log = Logger()
 
-    def __init__(self, eventloop=None):
+    def __init__(self, eventloop: Optional[SelectorEventLoop] = None):
+        if not eventloop:
+            _eventloop = get_event_loop()  # type: AbstractEventLoop
+            eventloop = cast(Optional[SelectorEventLoop], _eventloop)
 
-        if eventloop is None:
-            eventloop = get_event_loop()
+        # On Python 3.8+, asyncio.get_event_loop() on
+        # Windows was changed to return a ProactorEventLoop
+        if not isinstance(eventloop, SelectorEventLoop):
+            raise ValueError(
+                "eventloop is: {} which is not a subclass of "
+                "SelectorEventLoop".format(eventloop))
 
         self._asyncioEventloop = eventloop
```

Comment:
```suggestion
        self._asyncioEventloop = selected_eventloop  # type: SelectorEventLoop
```
Just to be explicit and overconstrain against the above logic to make sure it does what we want.  Maybe?

Reply:
I don't understand the sentence you wrote.

Reply:
It seems it might be good to explicitly state the type we want the `_asyncioEventLoop` attribute to have.  This avoids refactors of the logic preceding this assignment accidentally creating something else and unintentionally changing the type of the attribute.  It also causes mypy to check and confirm the result of the logic above really achieves the type we want.

Reply:
Thanks for the clarification.  I have incorporated your suggestion.